vowpalwabbit/csv_parser/src/parse_example_csv.cc

```diff
@@ -0,0 +1,364 @@
+        }
+        size_t end_idx = pair[1].length();
+        ns_value.insert(std::make_pair(ns, parseFloat(pair[1].data(), end_idx)));
+      }
+    }
+
+    // If no header present, will use empty features
+    if (all->csv_no_header && _header_fn.empty())
+    {
+      for (size_t i = 0; i < elements.size(); i++)
+      {
+        _header_fn.emplace_back(std::string());
+        _header_ns.emplace_back(std::string());
+      }
+    }
+
+    if (!_header_fn.empty() && elements.size() != _header_fn.size())
+    {
+      THROW(
+          "CSV line has " << elements.size() << " elements, but the header has " << _header_fn.size() << " elements!");
+    }
+    else if (!all->csv_no_header && _header_fn.empty())
+    {
+      for (size_t i = 0; i < elements.size(); i++)
+      {
+        if (all->csv_remove_quotes) { remove_quotation_marks(elements[i]); }
+
+        // Seperate the feature name and namespace from the header.
+        std::vector<VW::string_view> splitted = split(elements[i], all->csv_ns_separator);
+        VW::string_view feature_name;
+        VW::string_view ns;
+        if (splitted.size() == 1) { feature_name = elements[i]; }
+        else if (splitted.size() == 2)
+        {
+          ns = splitted[0];
+          feature_name = splitted[1];
+        }
+        else
+        {
+          THROW("Malformed header for feature name and namespace separator: " << elements[i]);
+        }
+        _header_fn.emplace_back(feature_name);
+        _header_ns.emplace_back(ns);
+      }
+    }
+    else
+    {
+      parse_example(all, ae, elements);
+    }
+  }
+}
+
+void parser::parse_example(VW::workspace* all, VW::example* ae, std::vector<VW::string_view> csv_line)
+{
+  if (label_list.empty() && !all->csv_label.empty())
+  {
+    VW::string_view csv_label(all->csv_label);
+    label_list = list_handler(csv_label, csv_line.size(), "label");
+  }
+  if (tag_list.empty() && !all->csv_tag.empty())
+  {
+    VW::string_view csv_tag(all->csv_tag);
+    tag_list = list_handler(csv_tag, csv_line.size(), "tag");
+    if (tag_list.size() > 1) { THROW("Can only have one tag at most per example!"); }
+  }
+
+  if (!label_list.empty()) { parse_label(all, ae, csv_line); }
+  if (!tag_list.empty()) { parse_tag(all, ae, csv_line); }
+
+  parse_namespaces(all, ae, csv_line);
+}
+
+void parser::parse_label(VW::workspace* all, VW::example* ae, std::vector<VW::string_view> csv_line)
+{
+  all->example_parser->lbl_parser.default_label(ae->l);
+
+  std::string label_content;
+  char label_part_separator = ' ';
+  if (all->csv_multilabels) { label_part_separator = ','; }
+
+  for (size_t i = 0; i < label_list.size(); i++)
+  {
+    VW::string_view label_content_part(csv_line[label_list[i]]);
+    if (all->csv_remove_quotes) { remove_quotation_marks(label_content_part); }
+    // Skip the empty cells
+    if (label_content_part.empty()) { continue; }
+    label_content += {label_content_part.begin(), label_content_part.end()};
```

Comment:
is this just appending the string_view to the string?
`label_content += label_content_part.to_string();` would be clearer

Reply:
> label_content += label_content_part.to_string();

![image](https://user-images.githubusercontent.com/43995067/170614825-c63b4db5-2c19-437a-86b8-dd4128644420.png)

Actually there's no to_string() for string_view...
And `{}` is one simple way to do this.
https://stackoverflow.com/a/67642605/14343335

Reply:
Ah, it looks like you're using a newer standard that has `std::string_view`. That version converts to a string via explicit constructor, so the `to_string` call is unnecessary. I'm vaguely surprised that this works on C++11, but I guess the nonstd header adds the missing string constructor for this case?